File: C++/src/Solution301_400/Solution383.cpp
//
// Created by windows on 2023/3/2.
//

#include "Solution383.h"
#include <iostream>
#include <vector>
#include <unordered_set>
#include <unordered_map>
#include <string>
using namespace std;
#include <unordered_map>

bool Solution383::canConstruct(string ransomNote, string magazine) {
    unordered_map<char, int> map;
    for (char c : magazine) {
        map[c]++;
    }
    for (char c : ransomNote) {
        if (!map.count(c)){
            return false;
        }
        if (map[c]-- <= 0) {
            return false;
        }
    }
    return true;
}

File: Java/src/main/java/QuestionBank/Solution301_400/Solution383.java
package QuestionBank.Solution301_400;

/**
 * @author windows
 */
public class Solution383 {
    public boolean canConstruct(String ransomNote, String magazine) {
        int[] chars = new int[26];
        for (int i = 0; i < magazine.length(); i++) {
            chars[magazine.charAt(i) - 'a']++;
        }
        for (int i = 0; i < ransomNote.length(); i++) {
            if (chars[ransomNote.charAt(i) - 'a'] == 0){
                return false;
            }else{
                chars[ransomNote.charAt(i) - 'a']--;
            }
        }
        return true;
    }


    public static void main(String[] args) {
        Solution383 solution383 = new Solution383();
    }
}